qsourceparser.py

```diff
@@ -0,0 +1,174 @@
+
+from collections import namedtuple
+import sys
+
+RawLine = namedtuple("RawLine", "linenumber raw")
+
+IndentedLine = namedtuple("IndentedLine", "linenumber raw indentation")
+
+Line = namedtuple("Line", "linenumber raw indentation operator arguments")
+
+
+def indentation_level(line):
+    level = 0
+    try:
+        while line.raw[level] == " ":
+            level += 1
+    except IndexError:
+        print("Invalid line %s on line %d" % (line.raw, line.linenumber))
+
+    assert level % 4 == 0, (
+        "Indentation level not a multiple of 4 on line %d" % line.linenumber)
+    return level / 4
+
+
+def no_comments(line):
+    position = line.find(';')
+    if position >= 0:
+        return line[:position]
+
+    return line
+
+
+def parse_lines(text):
+    raw_lines = [RawLine(i, no_comments(line)) for i, line in
+                 enumerate(text.splitlines())]
+
+    raw_lines = [line for line in raw_lines if
+                 line.raw and not line.raw.isspace()]
+
+
+    indented_lines = [IndentedLine(
+        line.linenumber, line.raw, indentation_level(line))
+        for line in raw_lines]
+
+    lines = []
+    for indented_line in indented_lines:
+        instruction = indented_line.raw.strip()
+        fields = instruction.split()
+        operator = fields[0]
+        arguments = fields[1:]
+        lines.append(Line(indented_line.linenumber,
+                          indented_line.raw,
+                          indented_line.indentation,
+                          operator,
+                          arguments))
+    return lines
+
+
+Block = namedtuple("Block", "head body")
+
+
+def parse_blocks(lines):
+    """Returns `[]Block` from `lines`."""
+    blocks = []
+    while len(lines) > 0:
+        head = lines.pop(0)
+
+        if len(lines) > 0:
+            next_line = lines[0]
+            if next_line.indentation == head.indentation:
+                blocks.append(Block(head, []))
+
+            elif next_line.indentation > head.indentation:
+                blocks.append(Block(head, parse_blocks(lines)))
+
+            elif next_line.indentation < head.indentation:
+                blocks.append(Block(head, []))
+                return blocks
+
+    return blocks
```

Comment:
@DanielJanzon Something like this is what I had in mind, so every call to `parse_blocks` parses the next indentation layer. A source with only one indentation level would become a list of Blocks with only the head filled in. The output of qcode_example.qc with an added loop instruction becomes

    olov@helium:~/proj/QuantumProgramming$ python qsourceparser.py 
    examples/qcode_example.qc 

    Line representation:

    Line(linenumber=1, raw='register q0[0]  ', indentation=0, operator='register', arguments=['q0[0]'])
    Line(linenumber=2, raw='register q1[1]', indentation=0, operator='register', arguments=['q1[1]'])
    Line(linenumber=3, raw='register q2[2,3]', indentation=0, operator='register', arguments=['q2[2,3]'])
    Line(linenumber=6, raw='macro hello a', indentation=0, operator='macro', arguments=['hello', 'a'])
    Line(linenumber=7, raw='    X a', indentation=1, operator='X', arguments=['a'])
    Line(linenumber=8, raw='    loop 5', indentation=1, operator='loop', arguments=['5'])
    Line(linenumber=9, raw='        H a', indentation=2, operator='H', arguments=['a'])
    Line(linenumber=11, raw='hello q0', indentation=0, operator='hello', arguments=['q0'])
    Line(linenumber=12, raw='hello q1', indentation=0, operator='hello', arguments=['q1'])
    Line(linenumber=15, raw='X q2', indentation=0, operator='X', arguments=['q2'])

    Block representation:

    Block: Line(linenumber=1, raw='register q0[0]  ', indentation=0, operator='register', arguments=['q0[0]'])
    Block: Line(linenumber=2, raw='register q1[1]', indentation=0, operator='register', arguments=['q1[1]'])
    Block: Line(linenumber=3, raw='register q2[2,3]', indentation=0, operator='register', arguments=['q2[2,3]'])
    Block: Line(linenumber=6, raw='macro hello a', indentation=0, operator='macro', arguments=['hello', 'a'])
        Block: Line(linenumber=7, raw='    X a', indentation=1, operator='X', arguments=['a'])
        Block: Line(linenumber=8, raw='    loop 5', indentation=1, operator='loop', arguments=['5'])
            Block: Line(linenumber=9, raw='        H a', indentation=2, operator='H', arguments=['a'])
    Block: Line(linenumber=11, raw='hello q0', indentation=0, operator='hello', arguments=['q0'])
    Block: Line(linenumber=12, raw='hello q1', indentation=0, operator='hello', arguments=['q1'])